x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/TransformDeprecationChecker.java

```diff
@@ -60,16 +76,54 @@ private void recursiveGetTransformsAndCollectDeprecations(
         request.setAllowNoResources(true);
 
         components.client().execute(GetTransformAction.INSTANCE, request, ActionListener.wrap(getTransformResponse -> {
+            CountDownLatch latch = new CountDownLatch(getTransformResponse.getTransformConfigurations().size());
             for (TransformConfig config : getTransformResponse.getTransformConfigurations()) {
                 issues.addAll(config.checkForDeprecations(components.xContentRegistry()));
+
+                ValidateTransformAction.Request validateTransformRequest = new ValidateTransformAction.Request(
+                    config,
+                    false,
+                    TimeValue.timeValueSeconds(30)
+                );
+                ActionListener<ValidateTransformAction.Response> validateTransformListener = new LatchedActionListener<>(
+                    ActionListener.wrap(validateTransformResponse -> {
+                        List<String> warningHeaders = components.client()
+                            .threadPool()
+                            .getThreadContext()
+                            .getResponseHeaders()
+                            .get("Warning");
+                        if (warningHeaders != null) {
+                            issues.addAll(
+                                warningHeaders.stream()
+                                    .map(warningHeader -> createDeprecationIssue(config.getId(), warningHeader))
+                                    .collect(toList())
+                            );
+                        }
+                    }, e -> { logger.warn("An exception occurred while gathering deprecation warnings for transform", e); }),
+                    latch
+                );
+                components.client().execute(ValidateTransformAction.INSTANCE, validateTransformRequest, validateTransformListener);
             }
+            latch.await();
```

Comment:
I am not sure it's acceptable to block this thread for up to 30 seconds.

An alternative approach that doesn't involve blocking a thread is what is used when force closing jobs: https://github.com/elastic/elasticsearch/blob/0699c9351f1439e246d408fd6538deafde4087b6/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java#L515-L517

An atomic counter is incremented after each response is processed.  Then the final listener is called in the listener that incremented the counter to the required number.  No thread is blocked waiting for all results.

Reply:
Done, but I used a more specialized class `CountDown` instead.